server/server.go

```diff
@@ -1930,9 +1953,33 @@ func (s *Server) createClient(conn net.Conn, ws *websocket) *client {
 	// Re-Grab lock
 	c.mu.Lock()
 
+	tlsRequired := ws == nil && info.TLSRequired
+	var pre []byte
+	// If we have both TLS and non-TLS allowed we need to see which
+	// one the client wants.
+	if opts.TLSConfig != nil && opts.AllowNonTLS {
+		pre = make([]byte, 4)
+		c.nc.SetReadDeadline(time.Now().Add(secondsToDuration(opts.TLSTimeout)))
+		n, _ := io.ReadFull(c.nc, pre[:])
+		c.nc.SetReadDeadline(time.Time{})
+		pre = pre[:n]
+		if n > 0 && pre[0] == 0x16 {
```

Comment:
Are we sure this is going to be the case for all versions or configuration of TLS? Meaning that, will that first byte always be 0x16 depending on wether it is TLS v1.2 or v1.3, or if client certificate need to be checked, etc?

Reply:
Yes its same for TLS 1.1, 1.2 and 1.3. Old SSL was 0x80 IIRC, do not plan to support that.

Reply:
Ok, looked a bit more at the TLS code and it expects the first byte of the first message to be `recordTypeHandshake = 22` (which is 0x16). I had missed in readHanshake() that it does go into readRecord() which will do the check for 0x16, and instead saw in readHanshake that there was a switch with multiple values, so I wasn't sure if there would be different codes depending on the configuration.